File: packages/ios-im-ui/Sources/FlareIMUI/Components/MessageContentView.swift
import SwiftUI

/// Context passed to every content renderer.
public struct FlareContentContext {
    public let isSelf: Bool
    public let previewMode: Bool
    public let senderName: String?
    public let mediaState: FlareMediaDownloadState?
    public let onMediaAction: ((FlareMessageContent) -> Void)?

    public init(
        isSelf: Bool,
        previewMode: Bool = false,
        senderName: String? = nil,
        mediaState: FlareMediaDownloadState? = nil,
        onMediaAction: ((FlareMessageContent) -> Void)? = nil
    ) {
        self.isSelf = isSelf
        self.previewMode = previewMode
        self.senderName = senderName
        self.mediaState = mediaState
        self.onMediaAction = onMediaAction
    }
}

public typealias FlareContentBuilder = (FlareMessageContent, FlareContentContext) -> AnyView

/// Registry for product content types (`vote`, `task`…). Built-in types are
/// rendered directly by ``MessageContentView``; register a builder to add or
/// override a type.
public enum FlareContentRegistry {
    private static var builders: [String: FlareContentBuilder] = [:]
    public static func register(_ type: String, _ builder: @escaping FlareContentBuilder) {
        builders[type] = builder
    }
    public static func unregister(_ type: String) { builders.removeValue(forKey: type) }
    public static func lookup(_ type: String) -> FlareContentBuilder? { builders[type] }
}

/// Content-type dispatcher — renders a message body by type. Spec:
/// Message/MessageContentView (`MessageContentView`).
///
/// Media without `onMediaAction` is still consumable: an image opens the kit ``ImagePreviewView``
/// full screen (with a download key only when `onMediaDownload` is given; in a ``MessageListView`` as the
/// conversation's gallery, `spec/image-gallery-vectors.json`), a video opens the kit
/// ``VideoPlayerView`` full screen and plays, and a voice message plays inside its body, one at a time
/// in a ``MessageListView``. With `onMediaAction` every media tap goes to the host, as before. Files
/// and locations always go to the host — a file tap goes to `onMediaAction`, else to `onOpenFile`
/// (which leaves the image, video and voice defaults on).
///
/// A poll's option and a task's checkbox are controls only with `onVote` (the option's index) and
/// `onTaskToggle` (the state the user asks for); without them the bodies are read-only.
///
/// Links (in a text body and on a link card) go to `onOpenLink`; without it the body opens a web
/// address with the platform opener after ``safeExternalURL(_:)``, and opens nothing for any other
/// scheme. A link card whose URL is not a web address has no tap at all.
public struct MessageContentView: View {
    private let content: FlareMessageContent
    private let ctx: FlareContentContext
    private let onMediaDownload: ((FlareMessageContent) -> Void)?
    private let onOpenFile: ((FlareFileContent) -> Void)?
    private let onOpenLink: ((String) -> Void)?
    private let onVote: ((Int) -> Void)?
    private let onTaskToggle: ((Bool) -> Void)?
    /// The list's media defaults; nil outside a list, where a media body owns its own.
    private var mediaSession: FlareMediaSession?
    /// The message this body belongs to: a voice message's playback identity, and where its pictures are in
    /// the timeline's gallery.
    private var messageId: String?
    /// The timeline's gallery; nil outside a list, where a picture opens alone.
    private var gallery: FlareImageGallerySource?
    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme
    @Environment(\.openURL) private var openURL
    @Environment(\.flareStrings) private var strings

    public init(
        content: FlareMessageContent,
        isSelf: Bool = false,
        previewMode: Bool = false,
        senderName: String? = nil,
        mediaState: FlareMediaDownloadState? = nil,
        onMediaAction: ((FlareMessageContent) -> Void)? = nil,
        onMediaDownload: ((FlareMessageContent) -> Void)? = nil,
        onOpenFile: ((FlareFileContent) -> Void)? = nil,
        onOpenLink: ((String) -> Void)? = nil,
        onVote: ((Int) -> Void)? = nil,
        onTaskToggle: ((Bool) -> Void)? = nil
    ) {
        self.content = content
        self.ctx = FlareContentContext(
            isSelf: isSelf, previewMode: previewMode, senderName: senderName,
            mediaState: mediaState, onMediaAction: onMediaAction)
        self.onMediaDownload = onMediaDownload
        self.onOpenFile = onOpenFile
        self.onOpenLink = onOpenLink
        self.onVote = onVote
        self.onTaskToggle = onTaskToggle
    }

    /// This body inside a timeline: `session` is the list's media defaults, `messageId` the message, `gallery` the
    /// timeline's pictures.
    func mediaDefaults(_ session: FlareMediaSession?, messageId: String?,
                       gallery: FlareImageGallerySource? = nil) -> MessageContentView {
        var copy = self
        copy.mediaSession = session
        copy.messageId = messageId
        copy.gallery = gallery
        return copy
    }

    /// Whether a tap on `content` uses the kit default: an image, album, video or voice message whose host
    /// passed no media handler.
    static func usesMediaDefaults(_ content: FlareMessageContent, hostHandles: Bool) -> Bool {
        !hostHandles && (content is FlareImageContent || content is FlareImageGroupContent || content is FlareVideoContent
            || content is FlareAudioContent)
    }

    public var body: some View {
        let colors = FlareColors.of(scheme, brand: flareBrandTheme)
        let foreground = ctx.isSelf ? colors.messageOutgoingForeground : colors.messageIncomingForeground
        if let custom = FlareContentRegistry.lookup(content.type) {
            custom(content, ctx)
        } else if mediaSession == nil, Self.usesMediaDefaults(content, hostHandles: ctx.onMediaAction != nil) {
            FlareOwnedMediaSession { session in
                framedBody(session, foreground: foreground)
            }
        } else {
            framedBody(mediaSession, foreground: foreground)
        }
    }

    private func framedBody(_ session: FlareMediaSession?, foreground: Color) -> some View {
        VStack(alignment: .leading) {
            canonicalBody(session)
            if let media = ctx.mediaState, media.isDownloading {
                ProgressView(value: Double(min(max(media.progressPct, 0), 100)), total: 100)
            }
        }
        .environment(\.flareMessageBodyForeground, foreground)
        .foregroundStyle(foreground)
    }

    private var action: (() -> Void)? {
        guard let handler = ctx.onMediaAction else { return nil }
        return { handler(content) }
    }

    /// The host's download handler for this body, as the preview's download key.
    private var download: (() -> Void)? {
        guard let onMediaDownload else { return nil }
        return { onMediaDownload(content) }
    }

    /// Opening an album tile: the host's media handler takes the whole album; without one the kit previews the
    /// tile's image — in a timeline, the conversation's gallery from it — which the host's download handler can save.
    private func albumOpener(_ album: FlareImageGroupContent, session: FlareMediaSession?) -> ((Int) -> Void)? {
        if let handler = ctx.onMediaAction { return { _ in handler(album) } }
        guard let session else { return nil }
        return { index in
            let image = album.images[index]
            session.open(image, messageId: messageId, index: index, gallery: gallery,
                         onDownload: onMediaDownload.map { download in { download(image) } })
        }
    }

    /// Opening a link card: the host's link handler, else the platform opener for a web address. A card
    /// whose URL is not one has no tap, so nothing looks actionable that cannot act.
    private func openCard(_ url: String) -> (() -> Void)? {
        if let onOpenLink { return { onOpenLink(url) } }
        guard case .system(let safe) = TextMessageView.linkTap(url, hostHandles: false) else { return nil }
        return { openURL(safe) }
    }

    @ViewBuilder private func canonicalBody(_ session: FlareMediaSession?) -> some View {
        switch content {
        case let c as FlareTextContent:
            TextMessageView(text: c.text, isSelf: ctx.isSelf, mentions: c.mentions, onLinkTap: onOpenLink)
        case let c as FlareRichTextContent:
            RichTextMessageView(docJson: c.docJson, plainText: c.plainText, title: c.title, isSelf: ctx.isSelf,
                                onLinkTap: onOpenLink)
        case let c as FlareEmojiContent:
            EmojiMessageView(emoji: c.emoji, onTap: action)
        case let c as FlareStickerContent:
            StickerMessageView(url: c.url, packageId: c.packageId, stickerId: c.stickerId,
                width: c.width, height: c.height, onTap: action)
        case let c as FlareImageContent:
            ImageMessageView(src: c.thumbnailURL ?? c.url, width: 240, height: 180, alt: c.alt,
                             onTap: action ?? session.map { session in
                                 { session.open(c, messageId: messageId, index: 0, gallery: gallery, onDownload: download) }
                             })
        case let c as FlareImageGroupContent:
            ImageGroupMessageView(images: c.images, description: c.description, isSelf: ctx.isSelf,
                                  onOpen: albumOpener(c, session: session))
        case let c as FlareVideoContent:
            VideoMessageView(poster: c.poster, duration: Self.duration(c.durationSec),
                             onPlay: action ?? session.map { session in { session.present(c) } })
        case let c as FlareAudioContent:
            if let action {
                VoiceMessageView(seconds: c.durationSec, onPlay: action)
            } else if let session {
                FlareVoiceMessageBody(content: c, id: messageId ?? c.url, playback: session.voice)
            } else {
                VoiceMessageView(seconds: c.durationSec)
            }
        case let c as FlareFileContent:
            FileMessageView(name: c.name, size: Self.bytes(c.sizeBytes),
                            onOpen: action ?? onOpenFile.map { open in { open(c) } })
        case let c as FlareLocationContent:
            LocationMessageView(title: c.name, address: c.address, onOpen: action)
        case let c as FlareCardContent:
            ContactMessageView(name: c.title, subtitle: c.subtitle, avatarUrl: c.imageURL, onOpen: action)
        case let c as FlareLinkCardContent:
            LinkCardMessageView(title: c.title, domain: c.url, thumb: c.imageURL, description: c.description,
                                onOpen: action ?? openCard(c.url))
        case let c as FlarePollContent:
            VoteMessageView(title: c.title, options: c.options.map { FlareVoteOption($0) },
                            onSelect: onVote.map { vote in { _, index in vote(index) } })
        case let c as FlareTaskContent:
            TaskMessageView(title: c.title, meta: c.detail, done: c.done,
                            onToggle: onTaskToggle.map { toggle in { toggle(!c.done) } })
        case let c as FlareCalendarContent:
            LinkCardMessageView(title: c.title, description: c.timeRange, onOpen: action,
                icon: AnyView(Image(systemName: "calendar")))
        case let c as FlareMiniAppContent:
            LinkCardMessageView(title: c.title, domain: c.appId, thumb: c.thumbnailUrl,
                description: c.description, onOpen: action)
        case let c as FlareAnnouncementContent:
            LinkCardMessageView(title: c.title, description: c.body, onOpen: action,
                icon: AnyView(Image(systemName: "megaphone")), descriptionMaxLines: nil)
        case let c as FlareNotificationContent:
            SystemMessageView(text: c.text)
        case let c as FlarePlaceholderContent:
            SystemMessageView(text: c.label)
        case let c as FlareGenericContent:
            // The kit's own one-line summary: the body's label when it has one, else the word for
            // its wire type. Wrapping the label in brackets drew a literal "[]" for a body without one.
            UnknownMessageView(contentType: c.type, summary: flareMessagePreviewText(c, strings: strings), isSelf: ctx.isSelf)
        default:
            UnknownMessageView(contentType: content.type, isSelf: ctx.isSelf)
        }
    }

    static func duration(_ seconds: Int) -> String {
        String(format: "%02d:%02d", seconds / 60, seconds % 60)
    }

    static func bytes(_ b: Int) -> String {
        if b < 1024 { return "\(b) B" }
        if b < 1024 * 1024 { return String(format: "%.1f KB", Double(b) / 1024) }
        if b < 1024 * 1024 * 1024 { return String(format: "%.1f MB", Double(b) / 1024 / 1024) }
        return String(format: "%.1f GB", Double(b) / 1024 / 1024 / 1024)
    }
}

/// A voice message played by the kit: ``VoiceMessageView`` over the timeline's voice playback.
struct FlareVoiceMessageBody: View {
    let content: FlareAudioContent
    let id: String
    @ObservedObject var playback: FlareVoicePlayback

    var body: some View {
        let active = playback.activeId == id
        let phase = active ? playback.phase : .idle
        VoiceMessageView(seconds: content.durationSec,
                         playing: phase == .playing,
                         elapsedSeconds: active ? Int(playback.elapsed) : nil,
                         failed: phase == .failed,
                         onPlay: { playback.toggle(id: id, url: content.url) })
    }
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/MessageContentView.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.VolumeUp
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.LocationOn
import androidx.compose.material.icons.outlined.Movie
import androidx.compose.material.icons.outlined.CalendarMonth
import androidx.compose.material.icons.outlined.Apps
import androidx.compose.material.icons.automirrored.outlined.Announcement
import androidx.compose.material.icons.rounded.PlayCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Context passed to every content renderer. */
data class FlareContentContext(
    val isSelf: Boolean,
    val previewMode: Boolean = false,
    val senderName: String? = null,
    val mediaState: FlareMediaDownloadState? = null,
    val onMediaAction: ((FlareMessageContent) -> Unit)? = null,
)

typealias FlareContentBuilder = @Composable (FlareMessageContent, FlareContentContext) -> Unit

/**
 * Registry for product content types (`vote`, `task`…). Built-in types are
 * rendered directly by [MessageContentView]; register a builder to add/override.
 */
object FlareContentRegistry {
    private val builders = mutableMapOf<String, FlareContentBuilder>()
    fun register(type: String, builder: FlareContentBuilder) { builders[type] = builder }
    fun unregister(type: String) { builders.remove(type) }
    fun lookup(type: String): FlareContentBuilder? = builders[type]
}

/**
 * Content-type dispatcher — renders a message body by type. Spec:
 * Message/MessageContentView (`MessageContentView`). Built-in bodies are also the standalone public components; media uses the package's Coil loader.
 *
 * Taps: [onMediaAction], when passed, receives every tap on a media body and the kit opens nothing
 * itself. Without it the kit consumes received media: an image opens the image preview, a video opens
 * the video player, and a voice message plays in its bubble (one at a time inside a [MessageList]); a
 * file goes to [onOpenFile], which opens it outside the app after the host's URL gate. Location and cards
 * stay host actions, reached through [onMediaAction].
 *
 * Downloads: [onMediaDownload] is the download key of the kit's image preview, with the picture on screen; without it
 * the preview has no download key. Inside a [MessageList] a picture opens the conversation's gallery instead, whose
 * key downloads each picture through the list's `onMediaDownload`.
 *
 * Links: a link in a text body and a link card go to [onOpenLink] with the address. Without it the kit opens
 * an http or https address that passes [safeExternalUrl] with the platform opener ([LocalUriHandler]) and
 * never opens any other scheme; a link card whose address cannot be opened is not tappable.
 */
@Composable
fun MessageContentView(
    content: FlareMessageContent,
    isSelf: Boolean = false,
    senderName: String? = null,
    mediaState: FlareMediaDownloadState? = null,
    onMediaAction: ((FlareMessageContent) -> Unit)? = null,
    onOpenFile: ((FlareFileContent) -> Unit)? = null,
    onOpenLink: ((String) -> Unit)? = null,
    /** A tapped poll option (its index); without it the poll is read-only. */
    onVote: ((Int) -> Unit)? = null,
    /** A tapped task checkbox (the done state asked for); without it the task is read-only. */
    onTaskToggle: ((Boolean) -> Unit)? = null,
    /** The image preview's download key, with the picture on screen; without it the preview has none. */
    onMediaDownload: ((FlareMessageContent) -> Unit)? = null,
) {
    val ctx = FlareContentContext(isSelf, senderName = senderName, mediaState = mediaState, onMediaAction = onMediaAction)
    val custom = FlareContentRegistry.lookup(content.type)
    if (custom != null) { custom(content, ctx); return }

    val openLink = rememberFlareLinkOpener(onOpenLink)
    val tap = flareContentTap(content, hasMediaHandler = onMediaAction != null, hasFileHandler = onOpenFile != null, hasLinkHandler = onOpenLink != null)
    val inherited = LocalFlareMediaHost.current
    // Outside a MessageList the body presents its own media.
    val media = when {
        tap !is FlareContentTap.PreviewImage && tap !is FlareContentTap.PreviewAlbumImage && tap !is FlareContentTap.PlayVideo &&
            tap !is FlareContentTap.PlayVoice -> inherited
        inherited != null -> inherited
        else -> rememberFlareMediaHost().also { FlareMediaOverlay(it) }
    }
    val messageKey = LocalFlareMessageKey.current
    val gallery = LocalFlareImageGallery.current
    val voiceKey = messageKey ?: (content as? FlareAudioContent)?.url.orEmpty()
    val action: (() -> Unit)? = when (tap) {
        null -> null
        FlareContentTap.Host -> { { onMediaAction?.invoke(content) } }
        is FlareContentTap.PreviewImage -> { {
            media?.present(flareImagePresentation(gallery, messageKey, 0, tap.src, onMediaDownload?.let { download -> { download(content) } }))
        } }
        // An album's tiles open their own images; the body reports which one (below).
        FlareContentTap.PreviewAlbumImage -> null
        is FlareContentTap.PlayVideo -> { { media?.present(FlareMediaPresentation.Video(tap.src)) } }
        is FlareContentTap.PlayVoice -> { { media?.voice?.toggle(voiceKey, tap.src) } }
        FlareContentTap.OpenFile -> { { (content as? FlareFileContent)?.let { onOpenFile?.invoke(it) } } }
        is FlareContentTap.OpenLink -> { { openLink(tap.url) } }
    }

    val colors = flareColors()
    val foreground = if (isSelf) colors.messageOutgoingForeground else colors.messageIncomingForeground
    FlareThemeProvider(colors = colors.copy(
        textPrimary = foreground, textSecondary = foreground,
        textTertiary = foreground.copy(alpha = 0.8f),
        primary = if (isSelf) foreground else colors.primary,
    )) {
        Column {
            when (content) {
                is FlareTextContent -> TextMessage(content.text, self = isSelf, onLinkTap = openLink, mentions = content.mentions)
                is FlareRichTextContent -> RichTextMessage(content.docJson, plainText = content.plainText, title = content.title,
                    self = isSelf, onLinkTap = openLink)
                is FlareEmojiContent -> EmojiMessage(content.emoji, onTap = action)
                is FlareStickerContent -> StickerMessage(url = content.url, packageId = content.packageId,
                    stickerId = content.stickerId, width = content.width, height = content.height, onTap = action)
                is FlareImageContent -> ImageMessage(src = content.thumbnailUrl ?: content.url,
                    maxWidth = 240, maxHeight = 240, alt = content.alt, onTap = action)
                is FlareImageGroupContent -> ImageGroupMessage(
                    content.images, description = content.description, self = isSelf,
                    onOpen = when (tap) {
                        FlareContentTap.Host -> { _ -> onMediaAction?.invoke(content) }
                        FlareContentTap.PreviewAlbumImage -> { index ->
                            val image = content.images[index]
                            val source = image.url.ifBlank { image.thumbnailUrl.orEmpty() }
                            media?.present(flareImagePresentation(gallery, messageKey, index, source, onMediaDownload?.let { download -> { download(image) } }))
                        }
                        else -> null
                    },
                )
                is FlareVideoContent -> VideoMessage(poster = content.poster, duration = duration(content.durationSec), onPlay = action)
                is FlareAudioContent -> {
                    val voice = media?.voice?.takeIf { tap is FlareContentTap.PlayVoice }
                    val phase = voice?.phaseOf(voiceKey) ?: FlareVoicePhase.Idle
                    val learntMs = voice?.durations?.get(voiceKey) ?: 0
                    VoiceMessage(
                        seconds = if (content.durationSec > 0) content.durationSec else (learntMs + 500) / 1000,
                        playing = phase == FlareVoicePhase.Preparing || phase == FlareVoicePhase.Playing,
                        onPlay = action,
                        elapsedSeconds = if (phase == FlareVoicePhase.Playing || phase == FlareVoicePhase.Paused) voice?.positionMs?.div(1000) ?: 0 else 0,
                        failed = phase == FlareVoicePhase.Failed,
                    )
                }
                is FlareFileContent -> FileMessage(name = content.name, size = bytes(content.sizeBytes), onOpen = action)
                is FlareLocationContent -> LocationMessage(title = content.name, address = content.address, onOpen = action)
                is FlareCardContent -> ContactMessage(name = content.title, subtitle = content.subtitle,
                    avatarUrl = content.imageUrl, onOpen = action)
                is FlareLinkCardContent -> LinkCardMessage(title = content.title, domain = content.url,
                    description = content.description, thumb = content.imageUrl, onOpen = action)
                is FlarePollContent -> VoteMessage(
                    title = content.title,
                    options = content.options.map { FlareVoteOption(it) },
                    onSelect = onVote?.let { vote -> { _, index -> vote(index) } },
                )
                is FlareTaskContent -> TaskMessage(
                    title = content.title,
                    meta = content.detail,
                    done = content.done,
                    onToggle = onTaskToggle?.let { toggle -> { toggle(!content.done) } },
                )
                is FlareCalendarContent -> LinkCardMessage(title = content.title, description = content.timeRange,
                    onOpen = action, icon = { Icon(Icons.Outlined.CalendarMonth, null, tint = foreground) })
                is FlareMiniAppContent -> LinkCardMessage(title = content.title, domain = content.appId,
                    description = content.description, thumb = content.thumbnailUrl, onOpen = action)
                is FlareAnnouncementContent -> LinkCardMessage(title = content.title, description = content.body,
                    descriptionMaxLines = Int.MAX_VALUE, onOpen = action,
                    icon = { Icon(Icons.AutoMirrored.Outlined.Announcement, null, tint = foreground) })
                is FlareNotificationContent -> SystemMessage(content.text)
                is FlarePlaceholderContent -> SystemMessage(content.label)
                // The kit's own one-line summary: the body's label when it has one, else the word for its
                // wire type. Wrapping the label in brackets drew a literal "[]" for a body without one.
                is FlareGenericContent -> UnknownMessage(contentType = content.type, summary = flareMessagePreviewText(content, flareStrings()), isSelf = isSelf)
                else -> UnknownMessage(contentType = content.type, isSelf = isSelf)
            }
            if (mediaState?.isDownloading == true) {
                androidx.compose.material3.LinearProgressIndicator(progress = { mediaState.progressPct.coerceIn(0, 100) / 100f })
            }
        }
    }
}

internal fun duration(seconds: Int): String = "%02d:%02d".format(seconds / 60, seconds % 60)

internal fun bytes(b: Int): String = when {
    b < 1024 -> "$b B"
    b < 1024 * 1024 -> "%.1f KB".format(b / 1024.0)
    b < 1024 * 1024 * 1024 -> "%.1f MB".format(b / 1024.0 / 1024)
    else -> "%.1f GB".format(b / 1024.0 / 1024 / 1024)
}
